File: frontend/src/shared/lib/permissions.ts
import type { User } from "@/types/api";

const RESOURCE_MAP: Record<string, string> = {
  dashboard: "reports",
  job_titles: "structure",
  audit_logs: "settings",
};

const DEFAULT_RULES: Record<string, Set<"read" | "write">> = {
  reports: new Set(["read"]),
  self_service: new Set(["read", "write"]),
};

export const ADMIN_ROLES = new Set(["system_admin", "admin", "hr_manager"]);

const normalizeCodes = (codes?: string[]) =>
  (Array.isArray(codes) ? codes : [])
    .map((code) => String(code || "").trim().toLowerCase())
    .filter(Boolean);

const hasPermissionCode = (codes: string[], resource: string, action: "read" | "write") => {
  if (codes.includes("*") || codes.includes("all")) return true;
  const candidates = [resource, `${resource}.${action}`, `${resource}.*`];
  return candidates.some((code) => codes.includes(code));
};

export const getResourceFromRoute = (pathname: string) => {
  if (pathname === "/") return "reports";
  if (pathname.startsWith("/reports")) return "reports";
  if (pathname.startsWith("/employees")) return "employees";
  if (pathname.startsWith("/attendance")) return "attendance";
  if (pathname.startsWith("/leaves")) return "leaves";
  if (pathname.startsWith("/structure")) return "structure";
  if (pathname.startsWith("/devices")) return "devices";
  if (pathname.startsWith("/device-command-center")) return "devices";
  if (pathname.startsWith("/job-titles")) return "structure";
  if (pathname.startsWith("/users")) return "users";
  if (pathname.startsWith("/settings")) return "settings";
  if (pathname.startsWith("/audit-logs")) return "settings";
  if (pathname.startsWith("/payroll")) return "payroll";
  if (pathname.startsWith("/recruitment")) return "recruitment";
  if (pathname.startsWith("/performance")) return "performance";
  if (pathname.startsWith("/training")) return "training";
  if (pathname.startsWith("/assets")) return "assets";
  if (pathname.startsWith("/self-service")) return "self_service";
  return null;
};

export const canAccessResource = (
  user: User | null | undefined,
  resource: string | null,
  action: "read" | "write" = "read",
) => {
  if (!resource) return true;
  if (!user) return false;
  if (resource === "self_service") return true;

  const mapped = RESOURCE_MAP[resource] || resource;

  // Use server-provided effective permissions first when available.
  const effectivePermissions = normalizeCodes(user.effective_permissions);
  if (effectivePermissions.length > 0) {
    return hasPermissionCode(effectivePermissions, mapped, action);
  }

  // Explicit permissions list has priority and should match backend behavior.
  const explicitPermissions = normalizeCodes(user.permissions);
  if (explicitPermissions.length > 0) {
    return hasPermissionCode(explicitPermissions, mapped, action);
  }

  const roleCode = String(user.role || "").trim().toLowerCase();
  if (ADMIN_ROLES.has(roleCode)) {
    return true;
  }

  // Role-based permissions from role details (if provided)
  const rolePermissions = normalizeCodes(
    (user as { rolePermissions?: string[] }).rolePermissions || user.role_permissions || [],
  );
  if (rolePermissions.length > 0) {
    return hasPermissionCode(rolePermissions, mapped, action);
  }

  // Fallback: minimum safe access (dashboard + self service only)
  const allowed = DEFAULT_RULES[mapped] || new Set<"read" | "write">();
  return allowed.has(action);
};

export const hasExactPermission = (
  user: User | null | undefined,
  permissionCode: string
) => {
  if (!user) return false;
  
  const roleCode = String(user.role || "").trim().toLowerCase();
  if (ADMIN_ROLES.has(roleCode)) return true;

  const codes = [
    ...normalizeCodes(user.effective_permissions),
    ...normalizeCodes(user.permissions),
    ...normalizeCodes(
      (user as { rolePermissions?: string[] }).rolePermissions || user.role_permissions || [],
    ),
  ];
  
  const normalizedReq = permissionCode.trim().toLowerCase();
  if (codes.includes("*") || codes.includes("all")) return true;
  return codes.includes(normalizedReq);
};

File: backend/apps/authentication/permissions.py
from __future__ import annotations

from typing import Iterable, Set

from rest_framework.permissions import BasePermission, SAFE_METHODS

ROLE_RULES = {
    "system_admin": {"default": {"read", "write"}},
    "admin": {"default": {"read", "write"}},
    "hr_manager": {"default": {"read", "write"}},
    "supervisor": {
        "default": {"read"},
        "leaves": {"read", "write"},
        "attendance": {"read"},
    },
    "employee": {
        "default": {"read"},
        "me": {"read", "write"},
        "notifications": {"read", "write"},
    },
}

RESOURCE_MAP = {
    "employee": "employees",
    "employees": "employees",
    "attendance": "attendance",
    "leave": "leaves",
    "leaves": "leaves",
    "device": "devices",
    "devices": "devices",
    "device-groups": "devices",
    "device-policies": "devices",
    "device-templates": "devices",
    "device-command-approvals": "devices",
    "device-firmware-rollouts": "devices",
    "device-backups": "devices",
    "device-command-center": "devices",
    "department": "structure",
    "departments": "structure",
    "jobtitle": "structure",
    "job-titles": "structure",
    "settings": "settings",
    "notification": "notifications",
    "notifications": "notifications",
    "notificationread": "notifications",
    "notification-read": "notifications",
    "auditlog": "settings",
    "audit-logs": "settings",
    "user": "users",
    "users": "users",
    "roles": "settings",
    "role": "settings",
    "permissions": "settings",
    "permission": "settings",
    "payrollrecord": "payroll",
    "payroll": "payroll",
    "recruitmentcandidate": "recruitment",
    "recruitment": "recruitment",
    "performancereview": "performance",
    "performance": "performance",
    "trainingrecord": "training",
    "training": "training",
    "asset": "assets",
    "assets": "assets",
    "dashboard": "reports",
    "analytics": "reports",
    "reports": "reports",
    "webauthncredential": "settings",
    "employeedocument": "employees",
    "employee-documents": "employees",
}


def is_admin_role(user) -> bool:
    role = str(getattr(user, "role", "") or "").strip().lower()
    return role in {"system_admin", "admin", "hr_manager"}


def normalize_resource(resource: str | None) -> str:
    key = str(resource or "default").strip().lower()
    return RESOURCE_MAP.get(key, key)


def _normalize_codes(codes: Iterable[str] | None) -> Set[str]:
    if not codes:
        return set()
    normalized = set()
    for code in codes:
        value = str(code or "").strip().lower()
        if value:
            normalized.add(value)
    return normalized


def get_explicit_permission_codes(user) -> Set[str]:
    explicit = (
        getattr(user, "permissions_list", None)
        or getattr(user, "permissions", None)
        or []
    )
    return _normalize_codes(explicit)


def get_role_permission_codes(user) -> Set[str]:
    roles_manager = getattr(user, "roles_list", None) or getattr(user, "roles", None)
    if roles_manager is None:
        return set()
    return _normalize_codes(roles_manager.values_list("permissions__code", flat=True))


def get_effective_permission_codes(user) -> Set[str]:
    explicit_codes = get_explicit_permission_codes(user)
    if explicit_codes:
        return explicit_codes
    return get_role_permission_codes(user)


def permission_mode_for_user(user) -> str:
    if get_explicit_permission_codes(user):
        return "explicit"
    if get_role_permission_codes(user):
        return "role"
    return "legacy"


def _matches_permission(codes: Set[str], resource: str, action: str) -> bool:
    if {"*", "all"} & codes:
        return True
    candidates = {
        resource,
        f"{resource}.{action}",
        f"{resource}.*",
    }
    return bool(candidates & codes)


def has_any_permission_code(user, required_codes: Iterable[str]) -> bool:
    codes = get_effective_permission_codes(user)
    needed = _normalize_codes(required_codes)
    if codes:
        if {"*", "all"} & codes:
            return True
        return bool(codes & needed)
    return is_admin_role(user)


class RolePermission(BasePermission):
    """
    Central access control layer for modular apps.
    """

    def has_permission(self, request, view):
        if not request.user or not request.user.is_authenticated:
            return False

        resource = getattr(view, "basename", None) or getattr(view, "resource_name", "default")
        resource = normalize_resource(resource)
        action = "read" if request.method in SAFE_METHODS else "write"

        effective_codes = get_effective_permission_codes(request.user)
        if effective_codes:
            return _matches_permission(effective_codes, resource, action)

        role = str(getattr(request.user, "role", "employee") or "employee").strip().lower()
        rules = ROLE_RULES.get(role, ROLE_RULES["employee"])
        allowed = rules.get(resource, rules.get("default", {"read"}))
        return action in allowed
